### PR TITLE
Ensure there is at least one routing hint when creating invoice

### PR DESCRIPTION
**Pull request by roeierez**
This PR do a bit of refactor of the part adding the routing hints to the invoice.
The motivation for this was a real case when a user had private channels with a node that is not the lsp (same case as changing lsp after some channels were opened).
Creating an invoice for that user resulted without routing hints. The invoice returned from greenlight without routing hints and we didn't add routing hints because our previous logic only added hints for the current lsp and the user still didn't have a channel with the current lsp...

The logic here has been simplified to this:
1. If we need to open a channel the routing hints are only the lsp with the zero channel id
2. If we don't need to open a channel we: 
- use all private channels of the node to create the routing hints array (one channel per peer).
- We add to the invoice only the routing hints for peers that don't exist already in the hints returned from gl.

In addition, before using the node api to create the invoice we check if the user has routing hint candidates (private channels) and if not we return an early error.

### CODE REVIEW

File: libs/sdk-core/src/breez_services.rs [hunk 1]
```diff
@@ -2052,33 +2051,30 @@ impl Receiver for PaymentReceiver {
...
-                    };
+            routing_hints = self.node_api.get_routing_hints().await?;
+        }
 
-                    break;
-                }
-            }
+        if routing_hints.is_empty() {
+            return Err(ReceivePaymentError::InvoiceNoRoutingHints {
+                err: "Must be connected to at list one peer".into(),
```
Comment (ok300) [inline libs/sdk-core/src/breez_services.rs:2076]:
```suggestion
                err: "Must be connected to at least one peer".into(),
```

File: libs/sdk-core/src/greenlight/node_api.rs [hunk 2]
```diff
@@ -1435,6 +1435,67 @@ impl NodeAPI for Greenlight {
...
+        let mut hints: Vec<RouteHint> = vec![];
+        let mut node_client = self.get_node_client().await?;
+        let peers = node_client
+            .list_peers(cln::ListpeersRequest::default())
+            .await?
+            .into_inner();
+
+        // For each peer look for a private channel and create a routing hint from it.
+        for peer in peers.peers {
+            if !peer.channels.is_empty() {
```
Comment (ok300) [inline libs/sdk-core/src/greenlight/node_api.rs:1450]: IMO no need to check for this, because if the list is empty, the iterator below will just not iterate.

File: libs/sdk-core/src/breez_services.rs [hunk 3]
```diff
@@ -2052,33 +2051,30 @@ impl Receiver for PaymentReceiver {
...
-
-                    short_channel_id = match hint {
-                        None => None,
-                        Some(hint) => {
-                            let scid = parse_short_channel_id(&hint)?;
-                            info!("Found channel ID: {scid:?} {active_channel:?}");
-                            Some(scid)
-                        }
-                    };
+            routing_hints = self.node_api.get_routing_hints().await?;
```
Comment (ok300) [inline libs/sdk-core/src/breez_services.rs:2071]: The log statement above is not relevant anymore, as this is not finding the channel ID anymore.

File: libs/sdk-core/src/greenlight/node_api.rs [hunk 4]
```diff
@@ -1435,6 +1435,67 @@ impl NodeAPI for Greenlight {
         debug!("send_custom_message returned status {:?}", resp.status);
         Ok(())
     }
+
+    // Gets the routing hints related to all private channels that the node has
+    async fn get_routing_hints(&self) -> NodeResult<Vec<RouteHint>> {
```
Comment (JssDWt) [inline libs/sdk-core/src/greenlight/node_api.rs:1440]: This can potentially return many routing hints. I think we'd want to select a maximum of 3 hints. Probably ordered by the liquidity available (if possible). 

Note that a route hint is (usually) not useful if the peer only has private channels as well. So you might want to select nodes with active public channels in the lightning graph.

**Review by JssDWt — COMMENTED** [review]
Adding one comment now.
There's probably some nits, which I'll ignore for now.
I still have to check the `add_routing_hints` logic, which has become a bit hard to follow. 
Otherwise looks good.

File: libs/sdk-core/src/greenlight/node_api.rs (hunk 4 above)
Comment (roeierez) [reply libs/sdk-core/src/greenlight/node_api.rs:1440]: Do you see any real issue by leaving the logic trivial?  I don't want to complicate it if not really necessary.

File: libs/sdk-core/src/greenlight/node_api.rs [hunk 5]
```diff
@@ -1435,6 +1435,67 @@ impl NodeAPI for Greenlight {
...
+        let mut hints: Vec<RouteHint> = vec![];
+        let mut node_client = self.get_node_client().await?;
+        let peers = node_client
+            .list_peers(cln::ListpeersRequest::default())
```
Comment (JssDWt) [inline libs/sdk-core/src/greenlight/node_api.rs:1444]: This should be `list_peer_channels`. `list_peers` might also returns channels, but that functionality is deprecated and will be removed soon. (is already removed if deprecated apis are not enabled)

File: libs/sdk-core/src/invoice.rs [hunk 6]
```diff
@@ -145,15 +145,20 @@ impl RouteHint {
     }
 }
 
-pub fn add_lsp_routing_hints(
+pub fn add_routing_hints(
     invoice: String,
-    include_route_hints: bool,
-    lsp_hint: Option<RouteHint>,
+    merge_with_existing: bool,
```
Comment (JssDWt) [inline libs/sdk-core/src/invoice.rs:150]: If there are route hints in the invoice already, and we don't have to replace the route hints because there is no channel needed, do we have to merge in more route hints? Can we keep the ones selected by greenlight and be done with it? Only add route hints if there are none in that case?

File: libs/sdk-core/src/greenlight/node_api.rs (hunk 4 above)
Comment (JssDWt) [reply libs/sdk-core/src/greenlight/node_api.rs:1440]: Not necessarily. Chances are high greenlight will already have hop hints in the invoice. Just a limit of 2 or 3 hints would be good.

**Review by JssDWt — COMMENTED** [review]
2 more comments

EDIT: that don't show up here...

File: libs/sdk-core/src/greenlight/node_api.rs (hunk 4 above)
Comment (dangeross) [reply libs/sdk-core/src/greenlight/node_api.rs:1440]: I think the issue is GL is not reliably adding the hop hints, especially when creating an invoice very soon after connecting

File: libs/sdk-core/src/greenlight/node_api.rs [hunk 7]
```diff
@@ -1435,6 +1435,67 @@ impl NodeAPI for Greenlight {
...
+            .await?
+            .into_inner();
+
+        // For each peer look for a private channel and create a routing hint from it.
+        for peer in peers.peers {
+            if !peer.channels.is_empty() {
+                let optional_opened_channel = peer
+                    .channels
+                    .iter()
+                    .find(|&c| c.state == cln::ChannelState::ChanneldNormal as i32);
```
Comment (dangeross) [inline libs/sdk-core/src/greenlight/node_api.rs:1454]: This should also check the channel is private?

File: libs/sdk-core/src/invoice.rs [hunk 8]
```diff
@@ -166,28 +171,35 @@ pub fn add_lsp_routing_hints(
     // We make sure the hint we add does not conflict with other hints.
     // The lsp hint takes priority so in case the lsp hop is already in one of the existing hints
     // We make sure not to include them in the new hints.
```
Comment (dangeross) [inline libs/sdk-core/src/invoice.rs:173]:
```suggestion
    // We make sure the route hints we want to add do not conflict with invoice route hints.
    // When merging route hints, only route hints are added that go through different nodes than ones in the invoice route hints.
    // Otherwise when not merging route hints, the invoice route hints are replaced by the provided route hints.
```

File: libs/sdk-core/src/breez_services.rs [hunk 9]
```diff
@@ -2096,57 +2092,23 @@ impl Receiver for PaymentReceiver {
...
-            info!("Routing hint added");
-            let signed_invoice_with_hint = self.node_api.sign_invoice(raw_invoice_with_hint)?;
-            info!("Signed invoice with hint = {}", signed_invoice_with_hint);
+        info!("Routing hint added");
+        let signed_invoice_with_hint = self.node_api.sign_invoice(raw_invoice_with_hint)?;
+        info!("Signed invoice with hint = {}", signed_invoice_with_hint);
 
-            parsed_invoice = parse_invoice(&signed_invoice_with_hint)?;
-        }
+        parsed_invoice = parse_invoice(&signed_invoice_with_hint)?;
```
Comment (dangeross) [inline libs/sdk-core/src/breez_services.rs:2111]: I would suggest trying to ascertain if the routing hints have changed and only signing a new invoice if its the case. Else we're always creating two invoices, even if the route hints are no different to the GL ones.

**Comment by ok300** [comment]
Fixed conflicts with `main`.

File: libs/sdk-core/src/invoice.rs [hunk 10]
```diff
@@ -167,28 +172,35 @@ pub fn add_lsp_routing_hints(
...
-        None => invoice.route_hints(),
-        Some(lsp_hint) => match include_route_hints {
+    let unique_hop_hints: Vec<lightning::routing::router::RouteHint> = match route_hints.len() {
+        0 => invoice.route_hints(),
+        _ => match merge_with_existing {
             true => {
-                let mut all_hints: Vec<lightning::routing::router::RouteHint> = invoice
-                    .route_hints()
-                    .into_iter()
+                let unique_to_add: Vec<&RouteHint> = route_hints
```
Comment (ok300) [inline libs/sdk-core/src/invoice.rs:179]: Maybe a better way to find `unique_to_add` is

```rust
let invoice_hints_hop_src_node_ids: Vec<String> = invoice
    .route_hints()
    .into_iter()
    .flat_map(|hint| hint.0)
    .map(|hop| hop.src_node_id.serialize().encode_hex::<String>())
    .collect();
let unique_to_add: Vec<&RouteHint> = route_hints
    .iter()
    .filter(|hint| {
        hint.hops
            .iter()
            .all(|hop| !invoice_hints_hop_src_node_ids.contains(&hop.src_node_id))
    })
    .collect();
```

Avoids the 4-level-deep nesting in the filter, which is error-prone.

File: libs/sdk-core/src/breez_services.rs (hunk 3 above)
Comment (roeierez) [reply libs/sdk-core/src/breez_services.rs:2071]: fixed

File: libs/sdk-core/src/breez_services.rs (hunk 1 above)
Comment (roeierez) [reply libs/sdk-core/src/breez_services.rs:2076]: fixed

File: libs/sdk-core/src/breez_services.rs (hunk 9 above)
Comment (roeierez) [reply libs/sdk-core/src/breez_services.rs:2111]: done

File: libs/sdk-core/src/greenlight/node_api.rs (hunk 4 above)
Comment (roeierez) [reply libs/sdk-core/src/greenlight/node_api.rs:1442]: Added limit

File: libs/sdk-core/src/greenlight/node_api.rs (hunk 2 above)
Comment (roeierez) [reply libs/sdk-core/src/greenlight/node_api.rs:1452]: Addressed

File: libs/sdk-core/src/invoice.rs (hunk 6 above)
Comment (roeierez) [reply libs/sdk-core/src/invoice.rs:151]: changed int the caller scope to only include lsp hint if exists.

File: libs/sdk-core/src/invoice.rs (hunk 10 above)
Comment (roeierez) [reply libs/sdk-core/src/invoice.rs:185]: Done, thank you.

**Review by JssDWt — APPROVED** [review]
Looks good.

**Review by ok300 — APPROVED** [review]

File: libs/sdk-core/src/breez_services.rs [hunk 11]
```diff
@@ -2096,56 +2078,68 @@ impl Receiver for PaymentReceiver {
...
+        // limit the hints to max 3 and extract the lsp one.
+        let optional_lsp_hint = Self::limit_and_extract_lsp_hint(&mut routing_hints, &lsp_info);
+
+        // We here check if we need to modify the invoice.
+        let optional_modified_invoice = match (
+            open_channel_needed,
+            optional_lsp_hint,
+            parsed_invoice.routing_hints.is_empty(),
+        ) {
+            // If we need to open a channel we only need to set the decicated lsp hint.
```
Comment (dangeross) [inline libs/sdk-core/src/breez_services.rs:2093]:
```suggestion
            // If we need to open a channel we only need to set the dedicated lsp hint.
```

File: libs/sdk-core/src/breez_services.rs [hunk 12]
```diff
@@ -2096,56 +2078,68 @@ impl Receiver for PaymentReceiver {
...
-                req.amount_msat,
-            )?;
-
-            info!("Routing hint added");
-            let signed_invoice_with_hint = self.node_api.sign_invoice(raw_invoice_with_hint)?;
-            info!("Signed invoice with hint = {}", signed_invoice_with_hint);
+            // In case we don't need to open a channel and we have a channel with our lsp then we only ensure it
+            // exists as part of the invoice routing hints (merging).
+            (false, Some(h), _) => {
+                // In case we need to open a channel and the invoice has no routing hints we replace them with ours.
```
Comment (dangeross) [inline libs/sdk-core/src/breez_services.rs:2117]:
```suggestion
```

File: libs/sdk-core/src/breez_services.rs [hunk 13]
```diff
@@ -2096,56 +2078,68 @@ impl Receiver for PaymentReceiver {
...
+                )?)
+            }
+            (_, _, _) => None,
+        };
 
-            parsed_invoice = parse_invoice(&signed_invoice_with_hint)?;
+        if let Some(raw_invoice) = optional_modified_invoice {
+            let signed_invoice = self.node_api.sign_invoice(raw_invoice)?;
+            info!("Signed invoice with hint = {}", signed_invoice);
+            parsed_invoice = parse_invoice(&signed_invoice)?;
```
Comment (dangeross) [inline libs/sdk-core/src/breez_services.rs:2142]: I think for the normal case (not opening channel, have channel with lsp, have GL lsp route hint), this will still result in 2 invoices every time

Comment (roeierez) [reply libs/sdk-core/src/breez_services.rs:2147]: Good catch. Fixed

File: libs/sdk-core/src/breez_services.rs (hunk 12 above)
Comment (roeierez) [reply libs/sdk-core/src/breez_services.rs]: Done

File: libs/sdk-core/src/breez_services.rs (hunk 11 above)
Comment (roeierez) [reply libs/sdk-core/src/breez_services.rs]: Done

**Review by dangeross — APPROVED** [review]
LGTM!
